### PR TITLE
DD4hep RelVal for Phase2: update geometry configuration and PyReleaseValidation 

### PR DESCRIPTION
**Pull request by fabiocos**
#### PR description:

This PR extends #30003 introducing DD4hep geometry configurations and corresponding RelVal test workflows for Phase2 scenarios. In the initial version of this PR only the geometry D49 configuration is explicitly added (while workflows are possibly built for any Phase2 scenario). If there is agreement on this strategy I may update the PR by adding configurations for any active scenario, in any case the ```generateGeometry.py``` script is instructed to produce whatever Phase2 DD4hep configuration required. I add a few fixes to the geometry dictionaries to keep in synchrony them with the latest version available (for 2021) and have the dd4hep modifier available for MTD (other sub-detectors will have to check).

#### PR validation:

Code compiles and produces the desired configurations, tested for 2021 and 2026D49 both as compressed and expanded python configuration. The workflow list shows workflows for all scenarios, but their building fails but for D49, as the geometry configurations are not provided for the time being.

### CODE REVIEW

**Comment by fabiocos** [comment]
please test

**Comment by fabiocos** [comment]
```
16:26 farmui01 1668> runTheMatrix.py -n --what upgrade | grep DD4hep | grep D49
23207.911 2026D49_DD4hep+SingleMuPt10_Eta2p85_GenSimHLBeamSpot+DigiTrigger+RecoGlobal+HARVESTGlobal 
23234.911 2026D49_DD4hep+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal 
23250.911 2026D49_DD4hep+ZMM_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal 
23407.911 2026D49PU_DD4hep+SingleMuPt10_Eta2p85_GenSimHLBeamSpot+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU 
23434.911 2026D49PU_DD4hep+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU 
23450.911 2026D49PU_DD4hep+ZMM_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU 
16:26 farmui01 1669> runTheMatrix.py -n --what upgrade | grep DD4hep | wc -l
90
```

File: Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py [hunk 1]
```diff
@@ -846,15 +846,27 @@ def condition(self, fragment, stepList, key, hasHarvest):
...
+        elif 'Phase2' in stepDict[step][k]['--era']:
+            dd4hepGeom="DD4hep"
+            dd4hepGeom+=stepDict[step][k]['--geometry']
+            stepDict[stepName][k] = merge([{'--geometry' : dd4hepGeom, '--procModifiers': 'dd4hep'}, stepDict[step][k]])
     def condition(self, fragment, stepList, key, hasHarvest):
-        return (fragment=='TTbar_13' or fragment=='ZMM_13' or fragment=='SingleMuPt10') and '2021' in key
+        if (fragment=='TTbar_13' or fragment=='ZMM_13' or fragment=='SingleMuPt10') and '2021' in key:
+            return True
+        if (fragment=='TTbar_14TeV' or fragment=='ZMM_14' or fragment=='SingleMuPt10') and '2026' in key:
+            return True
```
Comment (kpedro88) [inline Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:859]: this condition is not complete (doesn't return false when the condition is not met)
I prefer avoiding the redundant `if true return true` construction:
```python
        return ((fragment=='TTbar_13' or fragment=='ZMM_13' or fragment=='SingleMuPt10') and '2021' in key) \
            or ((fragment=='TTbar_14TeV' or fragment=='ZMM_14' or fragment=='SingleMuPt10') and '2026' in key)
```

**Comment by kpedro88** [comment]
@fabiocos many thanks for following up on the discussion in #31902! the implementation here looks generally reasonable to me.

However, this PR will fail the unit test, which expects every available geometry cff to be generated. Therefore, all the configs should be added here.

**Comment by fabiocos** [comment]
@kpedro88 I was waiting for a comment, as said in the description I find reasonable to have all the possible scenarios available. I will update the PR, there is indeed a value in having all the scenarios available, apart for the unit test: for instance for MTD different parts of the code are probed when running different scenarios, and if we want to ensure that DD4hep is really working we need to run all of them (or at least a selection of them).

File: Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py (hunk 1 above)
Comment (fabiocos) [reply Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:859]: right, I forgot the ```False``` condition, I'll update that

**Comment by fabiocos** [comment]
please abort

**Comment by fabiocos** [comment]
hold

**Comment by fabiocos** [comment]
```scram b runtests``` looks ok now, I had to refresh a few configurations to add the DD4hep-compatible ```_cff``` versions of ```_cfi``` fragments. I have also realized that I was missing the update for TTbar_14TeV and ZMM_14 due to appending "14" at the end of the GENSIM step label, now this is fixed:

```

Preparing to run 23207.911 SingleMuPt10+2026D49_DD4hep+SingleMuPt10_Eta2p85_GenSimHLBeamSpot+DigiTrigger+RecoGlobal+HARVESTGlobal

Preparing to run 23234.911 TTbar_14TeV+2026D49_DD4hep+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

Preparing to run 23250.911 ZMM_14+2026D49_DD4hep+ZMM_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal
# in: /afs/infn.it/ts/project/cms/users/cossutti/Timing/dd4hep/CMSSW_11_2_X_2020-10-26-2300/work dryRun for 'cd 23207.911_SingleMuPt10+2026D49_DD4hep+SingleMuPt10_Eta2p85_GenSimHLBeamSpot+DigiTrigger+RecoGlobal+HARVESTGlobal
 cmsDriver.py SingleMuPt10_Eta2p85_cfi  --conditions auto:phase2_realistic_T15 -n 10 --era Phase2C9 --eventcontent FEVTDEBUG --procModifiers dd4hep --relval 9000,100 -s GEN,SIM --datatier GEN-SIM --beamspot HLLHC --geometry DD4hepExtended2026D49 --fileout file:step1.root  > step1_SingleMuPt10+2026D49_DD4hep+SingleMuPt10_Eta2p85_GenSimHLBeamSpot+DigiTrigger+RecoGlobal+HARVESTGlobal.log  2>&1
 


# in: /afs/infn.it/ts/project/cms/users/cossutti/Timing/dd4hep/CMSSW_11_2_X_2020-10-26-2300/work dryRun for 'cd 23234.911_TTbar_14TeV+2026D49_DD4hep+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal
 cmsDriver.py TTbar_14TeV_TuneCP5_cfi  --conditions auto:phase2_realistic_T15 -n 10 --era Phase2C9 --eventcontent FEVTDEBUG --procModifiers dd4hep --relval 9000,100 -s GEN,SIM --datatier GEN-SIM --beamspot HLLHC14TeV --geometry DD4hepExtended2026D49 --fileout file:step1.root  > step1_TTbar_14TeV+2026D49_DD4hep+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log  2>&1
 

# in: /afs/infn.it/ts/project/cms/users/cossutti/Timing/dd4hep/CMSSW_11_2_X_2020-10-26-2300/work dryRun for 'cd 23207.911_SingleMuPt10+2026D49_DD4hep+SingleMuPt10_Eta2p85_GenSimHLBeamSpot+DigiTrigger+RecoGlobal+HARVESTGlobal
 cmsDriver.py step2  --conditions auto:phase2_realistic_T15 -n 10 --era Phase2C9 --eventcontent FEVTDEBUGHLT --procModifiers dd4hep -s DIGI:pdigi_valid,L1TrackTrigger,L1,DIGI2RAW,HLT:@fake2 --datatier GEN-SIM-DIGI-RAW --geometry DD4hepExtended2026D49 --filein  file:step1.root  --fileout file:step2.root  > step2_SingleMuPt10+2026D49_DD4hep+SingleMuPt10_Eta2p85_GenSimHLBeamSpot+DigiTrigger+RecoGlobal+HARVESTGlobal.log  2>&1
 

# in: /afs/infn.it/ts/project/cms/users/cossutti/Timing/dd4hep/CMSSW_11_2_X_2020-10-26-2300/work dryRun for 'cd 23250.911_ZMM_14+2026D49_DD4hep+ZMM_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal
 cmsDriver.py ZMM_14TeV_TuneCP5_cfi  --conditions auto:phase2_realistic_T15 -n 10 --era Phase2C9 --eventcontent FEVTDEBUG --procModifiers dd4hep --relval 18000,100 -s GEN,SIM --datatier GEN-SIM --beamspot HLLHC14TeV --geometry DD4hepExtended2026D49 --fileout file:step1.root  > step1_ZMM_14+2026D49_DD4hep+ZMM_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log  2>&1
```

**Comment by fabiocos** [comment]
please test

**Comment by fabiocos** [comment]
unhold

**Comment by kpedro88** [comment]
@fabiocos out of curiosity, is 23234.911 functional right now, or does it crash somewhere?

**Comment by fabiocos** [comment]
@kpedro88 as of this morning not even 2021 was passing step1 (while it was sometime ago), although the latest updates by @ianna should hopefully fix this. I now see that in the development I have introduced a stupid residual bug in a path, the configuration structure is ok but the xml path is not. When fixing manually that:

```
A fatal system signal has occurred: segmentation violation
The following is the call stack containing the origin of the signal.

Wed Oct 28 19:09:58 CET 2020
Thread 3 (Thread 0x7f676b0ea700 (LWP 17499)):
#0  0x00007f67a2ce3c3d in poll () from /lib64/libc.so.6
#1  0x00007f6796c9443f in full_read.constprop () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00007f6796c94b7c in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00007f6796c95a59 in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00007f676d9472cf in operator<<(std::ostream&, G4Material const*) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#6  0x00007f676d947629 in operator<<(std::ostream&, G4Material const&) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#7  0x00007f676d7c8ded in dd4hep::sim::Geant4Converter::handleMaterial(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, dd4hep::Material) const () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#8  0x00007f676d7cbccd in dd4hep::sim::Geant4Converter::handleVolume(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, TGeoVolume const*) const () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#9  0x00007f676d7c614f in void (anonymous namespace)::handle<dd4hep::sim::Geant4Converter, std::vector<dd4hep::Volume, std::allocator<dd4hep::Volume> >, void* (dd4hep::sim::Geant4Converter::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, TGeoVolume const*) const>(dd4hep::sim::Geant4Converter const*, std::vector<dd4hep::Volume, std::allocator<dd4hep::Volume> > const&, void* (dd4hep::sim::Geant4Converter::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, TGeoVolume const*) const) [clone .constprop.990] () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#10 0x00007f676d7d21ac in dd4hep::sim::Geant4Converter::create(dd4hep::DetElement) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#11 0x00007f676d589762 in cms::DDG4Builder::BuildGeometry(SensitiveDetectorCatalog&) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#12 0x00007f676d58a507 in DDDWorld::DDDWorld(DDCompactView const*, cms::DDCompactView const*, SensitiveDetectorCatalog&, int, bool, bool) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#13 0x00007f676d50ddf0 in RunManagerMT::initG4(DDCompactView const*, cms::DDCompactView const*, HepPDT::ParticleDataTable const*) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#14 0x00007f676d502461 in OscarMTMasterThread::OscarMTMasterThread(edm::ParameterSet const&)::{lambda()#1}::operator()() const [clone .isra.86] () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#15 0x00007f67a35b3d2f in std::execute_native_thread_routine (__p=0x7f6776b17de0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#16 0x00007f67a2fc6ea5 in start_thread () from /lib64/libpthread.so.0
#17 0x00007f67a2cee8dd in clone () from /lib64/libc.so.6
Thread 2 (Thread 0x7f677efff700 (LWP 17489)):
#0  0x00007f67a2fce1d9 in waitpid () from /lib64/libpthread.so.0
#1  0x00007f6796c93fd7 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00007f6796c94a9a in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00007f67a35b3d2f in std::execute_native_thread_routine (__p=0x7f67900de1a0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#4  0x00007f67a2fc6ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f67a2cee8dd in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f67a12d6540 (LWP 17441)):
#0  0x00007f67a2fcaa35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f67a35ae62c in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_11_1_0_pre7-slc7_amd64_gcc820/build/CMSSW_11_1_0_pre7-build/BUILD/slc7_amd64_gcc820/external/gcc/8.2.0-bcolbf/gcc-8.2.0/obj/x86_64-unknown-linux-gnu/libstdc++-v3/include/x86_64-unknown-linux-gnu/bits/gthr-default.h:864
#2  std::condition_variable::wait (this=<optimized out>, __lock=...) at ../../../../../libstdc++-v3/src/c++11/condition_variable.cc:53
#3  0x00007f676d501d1b in OscarMTMasterThread::beginRun(edm::EventSetup const&) const () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#4  0x00007f676d4eb3aa in OscarMTProducer::globalBeginRun(edm::Run const&, edm::EventSetup const&, OscarMTMasterThread const*) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#5  0x00007f676d4f5ac4 in edm::stream::ProducingModuleAdaptor<OscarMTProducer, edm::stream::EDProducerBase, edm::stream::EDProducerAdaptorBase>::doBeginRun(edm::RunTransitionInfo const&, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/biglib/slc7_amd64_gcc820/pluginSimulation.so
#6  0x00007f67a5853e94 in edm::WorkerT<edm::stream::EDProducerAdaptorBase>::implDoBegin(edm::RunTransitionInfo const&, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-10-26-2300/lib/slc7_amd64_gcc820/libFWCoreFramework.so
(...)
```

**Comment by kpedro88** [comment]
Okay, good to know there's some debugging work to do. This PR provides a good test setup.

**Comment by fabiocos** [comment]
please test

**Comment by cvuosalo** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade

**Comment by silviodonato** [comment]
+1

**Comment by chayanit** [comment]
+1

**Pull request merged**